# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmValType.h
# Commit: b5cb4960f59b
# Full Hash: b5cb4960f59bbae82963e9e58f5d550f05e19d7b
# Author: Julian Seward <jseward@acm.org>
# Date: 2023-12-16 09:51:16
# Regressor Bug: 1861973
# File Overlap Count: 1
# Description:
#   Bug 1868820 - Change the interpretation of js::wasm::PackedTypeCode::typeDef_ from signed 48-bit to unsigned 48-bit.  r=yury.
#   
#   Bug 1861973 tried to clean up signedness issues relating to
#   js::wasm::PackedTypeCode::typeDef_ on 32 bit targets.  As a side effect, on 64
#   bit targets, it caused that field to be regarded as a 48-bit signed field,
# ==============================================================================

diff -r 0cf7507475be -r b5cb4960f59b js/src/wasm/WasmValType.h
--- a/js/src/wasm/WasmValType.h	Fri Dec 15 17:26:12 2023 +0000
+++ b/js/src/wasm/WasmValType.h	Fri Dec 15 17:37:29 2023 +0000
@@ -89,19 +89,17 @@
     MOZ_ASSERT_IF(tc != AbstractTypeRefCode, typeDef == nullptr);
     MOZ_ASSERT_IF(tc == AbstractTypeRefCode, typeDef != nullptr);
 #if defined(JS_64BIT) && defined(DEBUG)
-    // Double check that `typeDef` points inside the "canonical" 48-bit
-    // address space by checking that the lower 48 bits sign extend back to
-    // the original.  This is necessary since we will only store the lowest 48
-    // bits of it, as noted above.  There's no equivalent check on 32 bit
+    // Double check that `typeDef` only has 48 significant bits, with the top
+    // 16 being zero.  This is necessary since we will only store the lowest
+    // 48 bits of it, as noted above.  There's no equivalent check on 32 bit
     // targets since we can store the whole pointer.
-    int64_t w = int64_t(typeDef);
-    w <<= (64 - TypeDefBits);
-    w >>= (64 - TypeDefBits);
-    MOZ_ASSERT(w == int64_t(typeDef));
+    static_assert(sizeof(int64_t) == sizeof(uintptr_t));
+    uint64_t w = (uint64_t)(uintptr_t)typeDef;
+    MOZ_ASSERT((w >> TypeDefBits) == 0);
 #endif
     PackedTypeCode ptc = {};
     ptc.typeCode_ = PackedRepr(tc);
-    ptc.typeDef_ = (uintptr_t)typeDef;
+    ptc.typeDef_ = (uint64_t)(uintptr_t)typeDef;
     ptc.nullable_ = isNullable;
     return ptc;
   }
@@ -149,17 +147,10 @@
 
   const TypeDef* typeDef() const {
     MOZ_ASSERT(isValid());
-#if defined(JS_64BIT)
-    // Reconstitute the pointer by sign-extending the lowest TypeDefBits bits.
-    static_assert(sizeof(int64_t) == sizeof(uintptr_t));
-    int64_t w = int64_t(typeDef_);
-    w <<= (64 - TypeDefBits);
-    w >>= (64 - TypeDefBits);
-    return (const TypeDef*)(uintptr_t)w;
-#else
-    // The pointer is stored exactly in the lowest 32 bits of `typeDef_`.
+    // On a 64-bit target, this reconstitutes the pointer by zero-extending
+    // the lowest TypeDefBits bits of `typeDef_`.  On a 32-bit target, the
+    // pointer is stored exactly in the lowest 32 bits of `typeDef_`.
     return (const TypeDef*)(uintptr_t)typeDef_;
-#endif
   }
 
   bool isNullable() const {
